design/design-doc-packages-v2.md

```diff
@@ -0,0 +1,1387 @@
+   following responsibilities:
+  - Setting up all RBAC described in the [Background](#background) section.
+  - Annotating all CRDs with the proper labels.
+  - Creating the `ServiceAccount` for the controller `Deployment`, binding it to
+    its `ClusterRole` and starting the controller (i.e. creating the
+    `Deployment`).
+  - Making any modifications, such as
+    [syncing](https://github.com/crossplane/crossplane/blob/6fc50822fbf11a7d31f8a9dabde5c8948c3b36ac/pkg/controller/packages/pkg/pkg.go#L696)
+    the `Secret` for the `ServiceAccount` that are required for running the
+    controller in [host aware
+    mode](https://github.com/crossplane/crossplane/blob/master/design/one-pager-host-aware-stack-manager.md).
+
+The process for a `PackageInstall` is very similar, but the packages using the
+templating controller have the additional step of first producing a
+`StackDefinition` in the install `Job`, then translating that to a `Package` in
+the [_`StackDefinition`
+controller_](https://github.com/crossplane/crossplane/blob/6fc50822fbf11a7d31f8a9dabde5c8948c3b36ac/pkg/controller/packages/templates/stackdefinition_controller.go#L53).
+
+## Implementation
+
+This following sections describe the changes to the current package manager that
+are required to support the goals listed above.
+
+- [Package Format](#package-format)
+- [API Changes](#api-changes)
+- [Package Installation Configuration](#package-installation-configuration)
+- [Dependency Management](#dependency-management)
+- [Installation](#installation)
+- [Scenarios](#scenarios)
+- [Phase 1](#phase-1)
+- [Phase 2](#phase-2)
+- [Phase 3](#phase-3)
+
+### Package Format
+
+TODO(hasheddan): This section is still a WIP
+
+A Crossplane package is an OCI image that contains a `.registry/` directory in
+its filesystem. While the image is really just serving as a file format, it has
+a few properties that make it advantageous:
+
+- Allows integration with existing OCI image registries and tooling.
+- Allows a package to also include a binary that it intends to be run after
+  install is complete (i.e. the controller).
+- Allows package metadata schema and filesystem format to change over time
+  without having to modify the package manager because the logic can be included
+  in the package image to allow it to "unpack" itself in a format that the
+  package manager understands.
+
+In reality, the second two properties are somewhat mutually exclusive.
+Currently, a package does not unpack itself, but rather has its `.registry/`
+directory copied into the filesystem of an unpacking container that provides the
+logic for translating the manifests into a format that the package manager
+understands. This means that the `ENTRYPOINT` of the package image can be
+invoked directly in the `Deployment` that is created as part of installation,
+but also means that the package author must conform to whatever "unpack" image
+is being used, rather than building it into the package image itself.
```

Comment:
I guess it depends on who you ask, but I believe your assessment is correct. As mentioned, the former is a bit more flexible from the package manager perspective because it just invokes the entrypoint. That being said, the environment that code ran in would need to be heavily sandboxed because it would essentially be executing arbitrary code